feat: enhance custom guardrail functionality with dynamic HTTP headers

agrattan0820 opened the pull request:
## Enhance custom guardrail functionality with dynamic HTTP headers

<!-- e.g. "Implement user authentication feature" -->

## Relevant issues

<!-- e.g. "Fixes #000" -->

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem

<img width="988" height="62" alt="image" src="https://github.com/user-attachments/assets/196fcc91-4a1c-47c1-b1c4-6e6aa40e4646" />

## Type

<!-- Select the type of Pull Request -->
<!-- Keep only the necessary ones -->

🆕 New Feature
✅ Test

## Changes

- Added support for passing custom HTTP headers to the guardrail API via the `get_guardrail_custom_headers` method.
- Updated GraySwan guardrail to utilize custom headers in API requests.
- Improved documentation for custom guardrail usage, including examples for dynamic header configuration.
- Refactored payload preparation to prioritize custom headers over dynamic body parameters and configuration values.
- Added comprehensive tests for custom header extraction and payload preparation logic.

agrattan0820 committed:
feat: enhance custom guardrail functionality with dynamic HTTP headers

- Added support for passing custom HTTP headers to the guardrail API via the `get_guardrail_custom_headers` method.
- Updated GraySwan guardrail to utilize custom headers in API requests.
- Improved documentation for custom guardrail usage, including examples for dynamic header configuration.
- Refactored payload preparation to prioritize custom headers over dynamic body parameters and configuration values.
- Added comprehensive tests for custom header extraction and payload preparation logic.

agrattan0820 committed:
refactor: improve S3 cache test for concurrent operations and update guardrail payload handling

- Enhanced the S3 cache test to verify that all expected keys are present during concurrent operations.
- Updated the GraySwan guardrail tests to ensure custom headers are ignored for configuration parameters, prioritizing dynamic body values instead.
- Removed outdated tests related to custom header merging that are no longer applicable.

krrishdholakia commented:
@agrattan0820 looks like your testing is failing 

<img width="1278" height="423" alt="Screenshot 2025-11-18 at 7 55 01 PM" src="https://github.com/user-attachments/assets/3285fd70-7229-49ab-963d-c2ebf530aaa0" />

agrattan0820 committed:
refactor: streamline GraySwan guardrail tests and enhance payload preparation

- Refactored test functions for better readability and consistency by applying uniform formatting.
- Updated the GraySwan guardrail's payload preparation to include an optional `custom_headers` parameter.
- Improved assertions in tests to ensure accurate verification of expected behavior regarding custom headers and dynamic body values.

agrattan0820 committed:
Merge branch 'main' into feat/guardrails-customer-headers

agrattan0820 committed:
refactor: simplify code formatting and improve readability in main.py

- Removed unnecessary line breaks and streamlined conditional expressions for better clarity.
- Consolidated multiple lines into single lines where appropriate to enhance code conciseness.
- Ensured consistent formatting across various sections of the file.
